### PR TITLE
Data validation

### PR DESCRIPTION
**Pull request by shengjiayuu**
I've add some data validation checks using Pandera to our data analysis. Please review it and let me know if there's anything else I can do. Thanks

### CODE REVIEW

**Comment by HrayrMuradyan** [comment]
Here are some comments:
- Why some columns have nullable = False and some nullable = True? How those were decided?
- Why to change the data types of the following two columns to float type? "Flight Distance", "Departure Delay in Minutes". Integer is more memory efficient than float. (Better to change int64 to int32 as we don't expect to have very large integers)
- Duplicate rows should be checked with the test dataset also, so, better to merge them before looking for duplicates.
- I think it's better to correct the column names before doing checks.
- I suggest moving Data Validation section before EDA.

That's it for now, I will go over it and make the changes!

File: dsci522_environment.yml [hunk 1]
```diff
@@ -10,4 +10,5 @@ dependencies:
   - scikit-learn=1.3.2
   - seaborn=0.13.2
   - conda-lock=2.5.7
-  - jupyterlab=4.0.5
\ No newline at end of file
+  - jupyterlab=4.0.5
+  - pandera=0.7.1
```
[the remark this answers is not in the document]
Comment (AzinPiran) [reply dsci522_environment.yml:14]: @shengjiayuu , did you also update conda-linux-64.lock after adding pandera?

[the remark this answers is not in the document]
Comment (cs-uche) [reply dsci522_environment.yml:14]: > @shengjiayuu , did you also update conda-linux-64.lock after adding pandera?

The pull request shows changes to only the `environment.yml` and the notebook. I would assume that the `conda-linux-64.lock` was not updated.

**Comment by shengjiayuu** [comment]
> Here are some comments:
> 
> * Why some columns have nullable = False and some nullable = True? How those were decided?
> * Why to change the data types of the following two columns to float type? "Flight Distance", "Departure Delay in Minutes". Integer is more memory efficient than float. (Better to change int64 to int32 as we don't expect to have very large integers)
> * Duplicate rows should be checked with the test dataset also, so, better to merge them before looking for duplicates.
> * I think it's better to correct the column names before doing checks.
> * I suggest moving Data Validation section before EDA.
> 
> That's it for now, I will go over it and make the changes! Before that, please, tell me what do you think about my suggestions. Thanks for the wonderful work!

Thanks so much for your suggestions! 
1.For deciding whether a column should have nullable=False or nullable=True, I am making the columns that must not have missing values, and where missing data would likely result in inaccurate analysis, as nullable=False. For columns with nullable=True, missing data can be tolerated, and it's okay to have some missing values, since in real life, data is not always available or applicable to every record.However, those decisions are made based on my personal understanding, so thanks for pointing out and feel free to change them if you think it's needed. 
2. For my understanding, If we set the column's data type to int32 or int64, missing values cannot be stored. To handle missing values, we would typically need to fill them with 0, and this might affect the accuracy of the data. If we use the float type, we can represent missing values with NaN, which will not mislead or distort the data.
3. I corrected the column names after doing the checks because I realized that the column names needed to be changed while doing the check.
4. Feel free to change anything you think are correct tho!

File: dsci522_environment.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (shengjiayuu) [reply dsci522_environment.yml:14]: > > @shengjiayuu , did you also update conda-linux-64.lock after adding pandera?
> 
> The pull request shows changes to only the `environment.yml` and the notebook. I would assume that the `conda-linux-64.lock` was not updated.

Thanks so much for the reminder!! So sorry, I did forget to update the lock file 😭 will update this later.

**Comment by HrayrMuradyan** [comment]
Here are some of the changes I made:

- Moved all library imports to the top
- Added data validation checklist
- Moved data validation to the top
- Went step by step through all the required steps
- Combined the dataset (train + test) to do overall check
- Did train test split from scratch (we didn't know how they splitted before)
- Improved the plot of the target variable
- The ordinal features were already encoded, scaled them by minmax scaling
- With the new split, logistic regression was not the best, so I tuned the hyperparameter max_depth of the decision tree.
- Got better result, concluded the analysis with decision trees

And some other stuff that maybe I am missing. 
Please go over the notebook (It seems to be finalized, but I haven't checked it yet because I want to take a little rest).
If you have some suggestions or questions please free to ask.

**Comment by HrayrMuradyan** [comment]
Hi @cs-uche , I believe It's totally fine to look to the class distribution of the whole dataset. Even if you split the data into train and test, you can and probably should check the class distribution of the test data to see if you have enough observations from each class to test your model on (if you have large sample and you are randomly splitting, no need to worry about it, as the class imbalance will be very similar for train and test datasets).

In other words, looking to the class distribution we don't affect the training process, so no data leakage is possible.

**Comment by HrayrMuradyan** [comment]
I think we can add the target variable distribution check for training data separately in the EDA section to see if random splitting worked fine.

**Comment by AzinPiran** [comment]
Hi everyone, thank you so much for your great work. Please also consider my suggestion for improvement :

Areas for Improvement:

1- For the data validation part we also need to check the correlations between the target/response variable and features/explanatory variables, relationships or patterns between features and the target variable, could provide deeper context.

2- We can remove unnecessary columns in the preprocessing section, we can add another group of columns like drop_features and add id in that section, not in the EDA section. 

3- The method contains some steps for preprocessing and I guess some of the parts are missed please update it from this :
The data was pre-processed by:

Removing unnecessary columns
Scaling numerical features
Encoding categorical features

to this :

The data was pre-processed using the following steps:

Removing unnecessary columns: if we plan to add this drop_features
Scaling numerical features using StandardScaler
Encoding categorical features with OneHotEncoder (ignoring unknown categories)
Scaling ordinal features using MinMaxScaler

4- I am not sure if this comment is correct :
image

5- It is better to add an explanation about how missing values were handled, we should clarify it.

6- In our analysis we did these steps:
Baseline Model
Comparison Models
Metrics Evaluation
Hyperparameter Tuning

but what do you think about these steps:
Initial Model Scoring
Preliminary Optimization
Select Top Model
Comprehensive Hyperparameter Tuning

Maybe if we do some Hyperparameter Tuning also for LR, that model works better than the decision tree. But we can do this step in the next version. 

Again thank you so much.

**Comment by HrayrMuradyan** [comment]
1. This is checked using the correlation matrix and we remove one of the features as it has too high correlation.
2. We can, I will change it.
3. I will add it
4. Good catch!
5. There are no missing values as we drop the column containing them, I think no need to clarify it.
6. You mean, you want to add these sections? I think the analysis part is already fine, we have a baseline, we have several models, we do hyperparameter tuning and choose the best model.
7. Logistic Regression doesn't have much room for improvement. It is a linear model and makes his best with default parameters. The only thing we can do is to add regularization, which will reduce the score instead of improving it.


Thanks for the suggestions! I will update the notebook and push it for your review

**Comment by HrayrMuradyan** [comment]
I have added the changes, please review it and merge it if there are no other comments. 
I added some notes about missing values as well to clarify that by removing the column we get rid of them.

File: dsci522_environment.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (shengjiayuu) [reply dsci522_environment.yml:14]: Do you guys know how to fix this? I updated the conda-linux-64.lock file, and after that, I used docker build to check if it works. However, I encountered this error. @cs-uche @AzinPiran 
Screen Shot 2024-11-30 at 12 20 32 PM

**Comment by AzinPiran** [comment]
1- But in the correlation matrix I cannot see our target column , it is just done for our numeric columns right? please let me know if I am wrong. Is there anywhere else we were checking them?

**Comment by HrayrMuradyan** [comment]
Oh, you are right! I will add that now

**Comment by cs-uche** [comment]
> 1- But in the correlation matrix I cannot see our target column , it is just done for our numeric columns right? please let me know if I am wrong. Is there anywhere else we were checking them?

Great catch @HrayrMuradyan plese review the [checklist ](https://github.com/UBC-MDS/airline-customer-satisfaction-predictor/issues/16#issue-2692588999), this relates to the second to last item. 

Also please compare the analysis and check off the items completed.

**Comment by AzinPiran** [comment]
perfect, thanks for adding these plots. I am sure, now everything is completed.

**Comment by cs-uche** [comment]
The analysis looks good, I'll go ahead and close the pull request.
